File: packages/core/src/server/ws-file-watch.ts
import { realpath, stat } from 'node:fs/promises';
import { WebSocket } from 'ws';
import type { ChatManager } from '../chat/index.js';
import type { DaemonEvent } from '@qlan-ro/mainframe-types';
import type { FileWatcherService } from '../files/file-watcher.js';
import { resolveAndValidatePath } from './routes/path-utils.js';
import { createChildLogger } from '../logger.js';

const log = createChildLogger('ws-file-watch');

/** Build a composite map key so the same relative path under different projects/chats never collides. */
export function compositeKey(requestedPath: string, projectId?: string, chatId?: string): string {
  return `${projectId ?? ''}|${chatId ?? ''}|${requestedPath}`;
}

/**
 * Resolve a client-supplied path to an absolute, containment-validated path.
 * Absolute paths pass through; relative paths require a projectId.
 *
 * When both projectId and chatId are supplied, validates that the chat belongs
 * to the claimed project — rejects on mismatch to prevent cross-project traversal.
 */
export function resolveSubscribePath(
  chats: ChatManager,
  requestedPath: string,
  projectId?: string,
  chatId?: string,
): string | null {
  if (requestedPath.startsWith('/')) {
    return requestedPath;
  }
  if (!projectId) {
    log.warn({ path: requestedPath }, 'subscribe:file rejected: relative path requires projectId');
    return null;
  }
  if (chatId) {
    const chatProjectId = chats.getChatProjectId(chatId);
    if (chatProjectId !== null && chatProjectId !== projectId) {
      log.warn(
        { path: requestedPath, projectId, chatId, chatProjectId },
        'subscribe:file rejected: chat does not belong to claimed projectId',
      );
      return null;
    }
  }
  const basePath = chatId ? chats.getEffectivePath(chatId) : chats.getProjectPath(projectId);
  if (!basePath) {
    log.warn({ path: requestedPath, projectId, chatId }, 'subscribe:file rejected: project/worktree not found');
    return null;
  }
  const validated = resolveAndValidatePath(basePath, requestedPath);
  if (!validated) {
    log.warn({ path: requestedPath, basePath }, 'subscribe:file rejected: path escapes project base');
    return null;
  }
  return validated;
}

/**
 * Per-client file-watch state.
 * Tracks which absolute paths this client is subscribed to and maps a
 * composite key (projectId|chatId|requestedPath) → resolvedPath so
 * the same relative filename under two different projects never collides.
 */
export class WsFileWatch {
  readonly fileSubscriptions = new Set<string>();
  /** composite(projectId|chatId|requestedPath) → resolvedPath */
  readonly requestedToResolved = new Map<string, string>();

  async subscribe(
    requestedPath: string,
    absolutePath: string,
    fileWatcher: FileWatcherService,
    ws: WebSocket,
    projectId?: string,
    chatId?: string,
  ): Promise<void> {
    let resolvedPath: string;
    try {
      resolvedPath = await realpath(absolutePath);
    } catch {
      log.warn({ path: absolutePath }, 'subscribe:file rejected: realpath failed (file may not exist)');
      return;
    }
    try {
      const s = await stat(resolvedPath);
      if (!s.isFile()) {
        log.warn({ path: resolvedPath }, 'subscribe:file rejected: not a regular file');
        return;
      }
    } catch (err) {
      log.warn({ err, path: resolvedPath }, 'subscribe:file rejected: stat failed');
      return;
    }

    if (!this.fileSubscriptions.has(resolvedPath)) {
      this.fileSubscriptions.add(resolvedPath);
      fileWatcher.subscribe(resolvedPath);
      log.debug({ path: resolvedPath }, 'client subscribed to file');
    }
    this.requestedToResolved.set(compositeKey(requestedPath, projectId, chatId), resolvedPath);

    if (ws.readyState === WebSocket.OPEN) {
      const ack: DaemonEvent = { type: 'subscribe:file:ack', requestedPath, resolvedPath };
      ws.send(JSON.stringify(ack));
    }
  }

  unsubscribe(requestedPath: string, fileWatcher: FileWatcherService, projectId?: string, chatId?: string): void {
    const key = compositeKey(requestedPath, projectId, chatId);
    const resolvedPath = this.requestedToResolved.get(key);
    if (!resolvedPath) return;
    this.requestedToResolved.delete(key);
    this.fileSubscriptions.delete(resolvedPath);
    fileWatcher.unsubscribe(resolvedPath);
    log.debug({ path: resolvedPath }, 'client unsubscribed from file');
  }

  unsubscribeAll(fileWatcher: FileWatcherService): void {
    for (const filePath of this.fileSubscriptions) {
      fileWatcher.unsubscribe(filePath);
    }
    this.requestedToResolved.clear();
    this.fileSubscriptions.clear();
  }
}

File: packages/core-rs/crates/mainframe-server/src/ws_file_watch.rs
//! Ported from `src/server/ws-file-watch.ts`.
//!
//! Per-connection file-watch bookkeeping (PER_ENTITY per CONCURRENCY.tsv — lives
//! inside the connection's own state, guarded by that connection's mutex) plus
//! `resolveSubscribePath`: resolving a client-supplied path to an absolute,
//! containment-validated path, with the chat→project ownership check.

use std::collections::{HashMap, HashSet};

use mainframe_db::DatabaseManager;
use mainframe_services::files::FileWatcherService;

use crate::path_utils::resolve_and_validate_path;

/// Composite map key so the same relative path under different projects/chats
/// never collides. Mirrors `compositeKey`.
pub fn composite_key(
    requested_path: &str,
    project_id: Option<&str>,
    chat_id: Option<&str>,
) -> String {
    format!(
        "{}|{}|{}",
        project_id.unwrap_or(""),
        chat_id.unwrap_or(""),
        requested_path
    )
}

/// Per-client file-watch state. Tracks the resolved absolute paths this client
/// watches and maps a composite key → resolved path.
#[derive(Default)]
pub struct WsFileWatch {
    file_subscriptions: HashSet<String>,
    requested_to_resolved: HashMap<String, String>,
}

impl WsFileWatch {
    /// Realpath + is-file check, then register the watch (once per resolved path)
    /// and record the composite mapping. Returns `(requestedPath, resolvedPath)`
    /// for the `subscribe:file:ack`, or `None` if the file is missing / not a
    /// regular file (the TS logs a warn and returns without acking).
    pub async fn subscribe(
        &mut self,
        requested_path: &str,
        absolute_path: &str,
        file_watcher: &FileWatcherService,
        project_id: Option<&str>,
        chat_id: Option<&str>,
    ) -> Option<(String, String)> {
        let resolved = match tokio::fs::canonicalize(absolute_path).await {
            Ok(p) => p,
            Err(_) => {
                tracing::warn!(
                    path = absolute_path,
                    "subscribe:file rejected: realpath failed"
                );
                return None;
            }
        };
        match tokio::fs::metadata(&resolved).await {
            Ok(meta) if meta.is_file() => {}
            Ok(_) => {
                tracing::warn!(path = %resolved.display(), "subscribe:file rejected: not a regular file");
                return None;
            }
            Err(err) => {
                tracing::warn!(%err, path = %resolved.display(), "subscribe:file rejected: stat failed");
                return None;
            }
        }
        let resolved = resolved.to_string_lossy().into_owned();

        if !self.file_subscriptions.contains(&resolved) {
            self.file_subscriptions.insert(resolved.clone());
            file_watcher.subscribe(&resolved);
        }
        self.requested_to_resolved.insert(
            composite_key(requested_path, project_id, chat_id),
            resolved.clone(),
        );
        Some((requested_path.to_string(), resolved))
    }

    /// Remove the watch registered under this composite key. No-op if unknown.
    pub fn unsubscribe(
        &mut self,
        requested_path: &str,
        file_watcher: &FileWatcherService,
        project_id: Option<&str>,
        chat_id: Option<&str>,
    ) {
        let key = composite_key(requested_path, project_id, chat_id);
        let Some(resolved) = self.requested_to_resolved.remove(&key) else {
            return;
        };
        self.file_subscriptions.remove(&resolved);
        file_watcher.unsubscribe(&resolved);
    }

    /// Drop every watch this client holds (on socket close).
    pub fn unsubscribe_all(&mut self, file_watcher: &FileWatcherService) {
        for path in self.file_subscriptions.drain() {
            file_watcher.unsubscribe(&path);
        }
        self.requested_to_resolved.clear();
    }
}

/// The base directory for a *relative* `subscribe:file`, with the chat→project
/// ownership check. Runs on the DB thread (sync repo access) — the caller then
/// containment-validates the requested path against this base. Mirrors the
/// non-absolute branch of `resolveSubscribePath`; returns `None` (reject) on a
/// missing projectId, an ownership mismatch, or an unresolvable base.
///
/// Absolute paths are handled by the caller (returned as-is, no base) exactly as
/// the TS `if (requestedPath.startsWith('/')) return requestedPath;` fast-path.
pub fn resolve_subscribe_base(
    db: &DatabaseManager,
    project_id: &str,
    chat_id: Option<&str>,
) -> Option<String> {
    if let Some(cid) = chat_id {
        // Only reject when the chat exists AND belongs to a different project
        // (TS: `chatProjectId !== null && chatProjectId !== projectId`).
        if let Ok(Some(chat)) = db.chats.get(cid)
            && chat.project_id != project_id
        {
            return None;
        }
    }
    match chat_id {
        Some(cid) => effective_path(db, cid),
        None => project_path(db, project_id),
    }
}

/// `ChatManager.getEffectivePath(chatId)` approximation: the chat's stored
/// worktree path if set, else its project's path.
fn effective_path(db: &DatabaseManager, chat_id: &str) -> Option<String> {
    let chat = db.chats.get(chat_id).ok().flatten()?;
    match chat.worktree_path {
        Some(worktree) => Some(worktree),
        None => project_path(db, &chat.project_id),
    }
}

/// `ChatManager.getProjectPath(projectId)`: the project's root path.
fn project_path(db: &DatabaseManager, project_id: &str) -> Option<String> {
    db.projects.get(project_id).ok().flatten().map(|p| p.path)
}

/// Containment-validate a relative `requested_path` against `base` (async
/// realpath), matching the `resolveAndValidatePath(base, requested)` tail of
/// `resolveSubscribePath`.
pub async fn validate_relative(base: &str, requested_path: &str) -> Option<String> {
    resolve_and_validate_path(base, requested_path).await
}

// PORT STATUS: src/server/ws-file-watch.ts (WsFileWatch + resolveSubscribePath)
// confidence: medium
// todos: 1
// notes: WsFileWatch is per-connection PER_ENTITY state (guarded by the
// connection mutex in websocket.rs). `resolveSubscribePath` is split: the
// db-dependent base resolution + ownership check (`resolve_subscribe_base`) runs
// on the DB thread; the async realpath containment check (`validate_relative`)
// runs on the caller task. TODO(port-phase4): `effective_path` approximates
// ChatManager.getEffectivePath by reading the chat's stored `worktree_path`; the
// full live-worktree validation lands with the ChatManager port.
